docs/proposals/predictive-horizontal-pod-autoscaling.md

```diff
@@ -0,0 +1,232 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Predictive Horizontal Pod Autoscaling
+
+**Author**: Matt McNaughton (@mattjmcnaughton)
+
+## Preface
+
+This document describes the addition of prediction to the
+[horizontal pod autoscaler] (../user-guide/horizontal-pod-autoscaler.md).
+This addition will decrease the amount of time pods spend operating outside of
+their target value for resource consumption.
+
+## Overview and Motivation
+
+In 1.1, Kubernetes added the experimental horizontal autoscaling of pods,
+ensuring the number of pods automatically adjusts as applications’ demands rise
+and fall. This autoscaling, described in detail in the [documentation]
+(../user-guide/horizontal-pod-autoscaler.md), works based on the following
+calculation:
+
+```
+TargetNumOfPods = ceil(sum(CurrentPodsResourceConsumption) / TargetResourceConsumption)
+```
+
+If the selected pods are operating outside of their target resource
+consumption, the autoscaler will destroy or create pods until all the selected
+pods are operating in accordance with target consumption.
+
+Overall, this method of autoscaling works well, yet it could
+be improved by accounting for the amount of time necessary
+to initialize a pod (i.e. for a pod to be ready to share the computational
+work). A pod’s initialization time can be lengthy if the pod must initialize
+a framework, download shard files from a
+network, register itself with some type of manager, etc.
+More specifically, if the newly created pods take a long time to
+initialize, the already existing pods will spend the entire
+initialization time operating outside of target
+resource consumption while they wait for the newly created pods to
+share the computational work. Depending on pod initialization time, and the
+danger of operating outside target resource consumption, it can be
+extremely detrimental to not account for pod initialization time.
+
+**Predictive horizontal pod autoscaling** addresses this problem.
+It accounts for pod initialization time by
+determining the target number of pods based on the predicted future pod resource
+consumption, instead of the current pod resource consumption. In other words,
+if it takes 10 minutes for a pod to initialize, than if it is 6:00pm and
```

Comment:
Great question - I mostly just choose 10 minutes to have a nice round number for the hypothetical example. But with respect to contributors to pod initialization time, we were thinking lengthy framework initialization time, the need to download state before the container can run, the need to register with some type of external service or manager that may not immediately respond, among other things.

I hope to do more work soon in terms of examining the distribution of pod initialization times and particularly examining the the uppermost initialization times.